src/lib/party.ts

```diff
@@ -772,13 +638,31 @@ export class Party {
     return this.body.publicKey;
   }
 
+  sign(privateKey?: PrivateKey) {
```

Comment:
Can we keep these data types purely functional? This should lead to fewer gotchas for people.

ie. returning a copy of this Party but with the authorization field set

Reply:
Hmm.. So currently, our model for building up the transaction involves actions which mutate the SmartContract's `this.self` party, e.g.
```ts
this.balance.addInPlace(100); // mutates party.body.balance_change
this.x.set(1); // mutates party.body.update.app_state
this.self.update.permissions.setValue(...);
```
I'm not so excited about breaking this flow by introducing a command that replaces the entire party:
```ts
this.self = this.self.sign();
```

So how about this: I could make the `.sign` method on `Party` purely functional, but add another `.sign` method on SmarContract which performs the mutation `this.self = this.self.sign()`?

Reply:
I started doing this
https://github.com/o1-labs/snarkyjs/pull/123/commits/65937625263c6e9f464cd83f8ebd9aec9d335728
https://github.com/o1-labs/snarkyjs/pull/123/commits/7753a674552cbf9e1585a82eedce620ad3b6d336

I'm still having doubts that this is a good approach